webauthn-authenticator-rs/src/cbor/make_credential.rs

```diff
@@ -0,0 +1,215 @@
+        });
+
+        pin_uv_auth_param.map(|p| {
+            keys.insert(0x08, Value::Bytes(p));
+        });
+
+        pin_uv_auth_proto.map(|p| {
+            keys.insert(0x09, Value::Integer(p.into()));
+        });
+        MakeCredentialRequestRawDict { keys }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::cbor::make_credential::*;
+    use base64urlsafedata::Base64UrlSafeData;
+    use serde_cbor::{from_slice, to_vec, Value};
+    use webauthn_rs_proto::{PubKeyCredParams, RelyingParty, User};
+
+    #[test]
+    fn make_credential() {
+        let _ = tracing_subscriber::fmt().try_init();
+
+        // let map: BTreeMap<Value, Value> = BTreeMap::new();
+
+        // clientDataHash 0x01
+        // rp 0x02
+        // user 0x03
+        // pubKeyCredParams 0x04
+        // excludeList 0x05
+        // extensions 0x06
+        // options 0x07
+        // pinUvAuthParam 0x08
+        // pinUvAuthProtocol 0x09
+        // enterpriseAttestation 0x0A
+
+        /*
+        Dec 28 18:41:01.160  INFO webauthn_authenticator_rs::nfc::apdu::tests: got APDU Value response: Ok(Map(
+
+        Integer(1): Bytes([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
+        Integer(1): Bytes([104, 113, 52, 150, 130, 34, 236, 23, 32, 46, 66, 80, 95, 142, 210, 177, 106, 226, 47, 22, 187, 5, 184, 140, 37, 219, 158, 96, 38, 69, 241, 65]),
+        Integer(2): Map({
+            Text("id"): Text("test"),
+            Text("name"): Text("test")
+        }),
+        Integer(2): Map({
+            Text("id"): Text("test.ctap"),
+            Text("name"): Text("test.ctap")
+        }),
+        Integer(3): Map({
+            Text("id"): Bytes([116, 101, 115, 116]),
+            Text("name"): Text("test"),
+            Text("displayName"): Text("test")
+        }),
+        Integer(3): Map({
+            Text("id"): Bytes([43, 102, 137, 187, 24, 244, 22, 159, 6, 159, 188, 223, 80, 203, 110, 163, 198, 10, 134, 27, 154, 123, 99, 148, 105, 131, 224, 181, 119, 183, 140, 112]),
+            Text("name"): Text("testctap@ctap.com"),
+            Text("displayName"): Text("Test Ctap")
+        }),
+        Integer(4): Array([
+            Map({Text("alg"): Integer(-7), Text("type"): Text("public-key")})
+        ])
+        Integer(4): Array([
+            Map({Text("alg"): Integer(-7), Text("type"): Text("public-key")}),
+            Map({Text("alg"): Integer(-257), Text("type"): Text("public-key")}),
+            Map({Text("alg"): Integer(-37), Text("type"): Text("public-key")})
+        ]),
+        // I think this is incorrect?
+        Integer(6): Map({Text("hmac-secret"): Bool(true)}),
+        // May need to set uv false?
+        Integer(7): Map({Text("rk"): Bool(true)}),
+        // Not needed
+        Integer(8): Bytes([252, 67, 170, 164, 17, 217, 72, 204, 108, 55, 6, 139, 141, 161, 213, 8]),
+        // Not needed
+        Integer(9): Integer(1)}))
+        */
+
+        // why is the a prepended 0x01?
```

Comment:
There is a reason for it buried in the ctap spec, and I forgot where ...

Reply:
Yeah, this was something I copied from existing code. The reason is it's the error code :)